packages/grafana-runtime/src/services/pluginExtensions/utils.tsx

```diff
@@ -0,0 +1,98 @@
+import {
+  ComponentTypeWithExtensionMeta,
+  type PluginExtension,
+  type PluginExtensionComponent,
+  type PluginExtensionLink,
+  PluginExtensionTypes,
+} from '@grafana/data';
+
+export function isPluginExtensionLink(extension: PluginExtension | undefined): extension is PluginExtensionLink {
+  if (!extension) {
+    return false;
+  }
+  return extension.type === PluginExtensionTypes.link && ('path' in extension || 'onClick' in extension);
+}
+
+export function isPluginExtensionComponent(
+  extension: PluginExtension | undefined
+): extension is PluginExtensionComponent {
+  if (!extension) {
+    return false;
+  }
+  return extension.type === PluginExtensionTypes.component && 'component' in extension;
+}
+
+export function getLimitedComponents<Props>({
+  props,
+  components,
+  limit,
+  pluginIdPatterns,
+}: {
+  props: Props;
+  components: Array<ComponentTypeWithExtensionMeta<Props>>;
+  limit?: number;
+  pluginIdPatterns?: string[];
+}) {
+  if (!components.length) {
+    return null;
+  }
+
+  const renderedComponents: Array<ComponentTypeWithExtensionMeta<Props>> = [];
+
+  for (const Component of components) {
+    const { meta } = Component;
+
+    if (pluginIdPatterns && !createRegexFromPluginIdPatterns(pluginIdPatterns).test(meta.pluginId)) {
+      continue;
+    }
+
+    // If no limit is provided, return all components
+    if (limit === undefined) {
+      renderedComponents.push(Component);
+      continue;
+    }
+
+    if ((Component as Function)(props) != null) {
+      renderedComponents.push(Component);
+    }
+
+    // Stop if we've reached the limit
+    if (renderedComponents.length >= limit) {
+      break;
+    }
+  }
+
+  return renderedComponents;
+}
+
+export function renderLimitedComponents<Props>({
+  props,
+  components,
+  limit,
+  pluginIdPatterns,
+}: {
+  props: Props;
+  components: Array<ComponentTypeWithExtensionMeta<Props>>;
+  limit?: number;
+  pluginIdPatterns?: string[];
+}) {
+  const limitedComponents = getLimitedComponents({ props, components, limit, pluginIdPatterns });
+
+  if (!limitedComponents?.length) {
+    return null;
+  }
+
+  return (
+    <>
+      {limitedComponents.map((Component, index) => (
+        <Component key={index} {...props} />
```

Comment:
Maybe we should use something more unique then the `index` here, can we use the `meta` from the Component?

Reply:
Thanks, that's a good point @hugohaggmark 🙏 Fixed in https://github.com/grafana/grafana/pull/101331/commits/fb01e421fa425908a01c32b97982b5b2e9d434d1.